### ISSUE TITLE
4f268ad seems to have changed the behaviour of reverse lookups.

### ISSUE THREAD

**Issue by bmcustodio** [coredns/coredns#3686]
**What happened**:

While facing and investigating an [issue](https://github.com/coreos/etcd-operator/issues/2160#issuecomment-588400020) with `etcd-operator`, I found out that the behaviour of reverse lookups has changed between versions of CoreDNS. I was able to trace it to CoreDNS 1.6.0, and I `git bisect`ed my way to 4f268ad:

Before 4f268ad:

```shell
$ dig @127.0.0.1 -x 10.150.206.170

; <<>> DiG 9.10.6 <<>> @127.0.0.1 -x 10.150.206.170
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 22482
;; flags: qr aa rd; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 1
;; WARNING: recursion requested but not available

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
;; QUESTION SECTION:
;170.206.150.10.in-addr.arpa.   IN      PTR

;; ANSWER SECTION:
170.206.150.10.in-addr.arpa. 4  IN      PTR     cilium-etcd-hrb8cstc6f.cilium-etcd.cilium.svc.cluster.local.

;; Query time: 0 msec
;; SERVER: 127.0.0.1#53(127.0.0.1)
;; WHEN: Wed Feb 19 19:46:58 WET 2020
;; MSG SIZE  rcvd: 156
```

After 4f268ad:

```shell
dig @127.0.0.1 -x 10.150.206.170

; <<>> DiG 9.10.6 <<>> @127.0.0.1 -x 10.150.206.170
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 30465
;; flags: qr aa rd; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 1
;; WARNING: recursion requested but not available

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
;; QUESTION SECTION:
;170.206.150.10.in-addr.arpa.   IN      PTR

;; ANSWER SECTION:
170.206.150.10.in-addr.arpa. 5  IN      PTR     10-150-206-170.cilium-etcd-client.cilium.svc.cluster.local.

;; Query time: 0 msec
;; SERVER: 127.0.0.1#53(127.0.0.1)
;; WHEN: Wed Feb 19 19:47:18 WET 2020
;; MSG SIZE  rcvd: 155
```

I'd say the `client-go` bump (or the bump of some other dependency) caused the behaviour of reverse lookups to change. This makes `etcd-operator` (which seems to rely on reverse lookups) unable to properly create `etcd` clusters, and possibly breaks other applications that rely on reverse lookups.

**What you expected to happen**:

I expected things to just work, which means I didn't expect anything "special" to happen—which in its turn probably means I'd expect the behaviour of reverse lookups not to have change between versions.

**How to reproduce it (as minimally and precisely as possible)**:

This happened while trying to install [Cilium](https://github.com/cilium/cilium) on an AWS EKS (1.14.9) cluster running CoreDNS 1.6.6, but I think that it would be enough to install `etcd-operator`, create an `EtcdCluster` resource and wait for at least one pod to be created. Then, perform a reverse lookup against the IP of said pod as indicated above.

**Anything else we need to know?**:

I don't think so.

**Environment**:

- the version of CoreDNS: 1.6.0+.
- Corefile:

   ```Corefile
    .:53 {
        errors
        health
        kubernetes cluster.local in-addr.arpa ip6.arpa {
          pods insecure
          upstream
          fallthrough in-addr.arpa ip6.arpa
        }
        prometheus :9153
        forward . /etc/resolv.conf
        cache 30
        loop
        reload
        loadbalance
    }
   ```

- logs, if applicable: N/A.
- OS (e.g: `cat /etc/os-release`): Amazon Linux 2 (https://github.com/awslabs/amazon-eks-ami/).
- Others: N/A.

**Comment by chrisohaver** [comment]
`cilium-etcd-hrb8cstc6f` looks like a Pod name... which suggests that kubernetes plugin option `endpoint_pod_names` may have been enabled in the first example.

When `endpoint_pod_names` is not enabled, then the hostname of the Pod is used, and if that is not defined (in the ~Pod spec~ Endpoint API), then a dashed ip address is used (as in the second example).

**Edit**:  Actually I'm wrong about that... the name used in endpoint records always prefers the hostname if defined...  the order of preference is...

hostname -> pod name (if endpoint_pod_names is enabled) -> dashed ip

**Comment by chrisohaver** [comment]
Can you share/point to the etcd-operator Deployment yaml here (or however else it is deployed?)

**Comment by chrisohaver** [comment]
is it more or less this one: https://github.com/coreos/etcd-operator/blob/master/example/deployment.yaml ?

**Comment by bmcustodio** [comment]
@chrisohaver thanks for having a look! I can't reach the cluster at the moment, but judging by https://github.com/cilium/cilium-etcd-operator/blob/fdf1993dded89db77742916a9e55ef898a4e5ad0/etcd-operator/descriptors.go#L38 I'd say it's very similar to what you've linked.

I'd say it would be enough for you to create a basic `EtcdCluster` resource following the examples and perform a reverse lookup against one of the pods with, say, CoreDNS 1.6.7 and observe the current behaviour (let's call it "bad"). Then, replace 1.6.7 with 1.5.2, perform a reverse lookup against the same pod, and observe the old ("good") behaviour.

**Comment by bmcustodio** [comment]
> the order of preference is...

Judging by the fact that the behaviour changed in a commit that bumps dependencies but changes nothing otherwise, I'd say this order of preference is (or wasn't?) being enforced by CoreDNS itself - or am I reading this wrong?

**Comment by benfuu** [comment]
The etcd operator will define the hostname as part of the pod spec:
https://github.com/coreos/etcd-operator/blob/b2a63ce924602be64a156361f3ee85498843026a/pkg/util/k8sutil/k8sutil.go#L399-L400

I'm thinking maybe the issue lies in the fact that etcd-operator creates two services, one headless one called `etcd-cluster` (headless without clusterIP for the peers) and one called `etcd-cluster-client` (normal clusterIP service for clients).

Both services have their selectors referencing the pods, which have their hostname defined as `etcd-cluster`. The [k8s DNS spec](https://github.com/kubernetes/dns/blob/master/docs/specification.md#24---records-for-a-headless-service) specifies that for headless services, there should be an A record mapping the hostname (defined in the pod spec) to the endpoint IP (`hostname -> pod/endpoint IP`). For normal clusterIP services, there should be an A record mapping the service name to the cluster IP. However, it looks like CoreDNS is creating an additional A record for a clusterIP service endpoint in the dashed-ip form that points to the endpoint IP as @chrisohaver mentioned (`dashed-ip-form-of-endpoint -> pod/endpoint IP`). This means that there are two A records that will point to the endpoint IP, which makes sense why I am getting both the hostname and the dashed-ip form when doing a reverse lookup (the PTR record changes or there are multiple PTR records and a different one is returned each time a reverse lookup is done).

Is this the expected behavior? It seems that CoreDNS isn't doing anything wrong by creating an additional A record with the dashed-ip form even for non-headless services that points to the endpoint IP, but this results in reverse lookups being ambiguous. Similarly, the etcd-operator is correctly setting the hostname for the pods and expects all reverse lookups for the pod IP to be that hostname, not the dashed-ip form of the endpoint.

**Comment by bmcustodio** [comment]
I've enabled `endpoint_pod_names` in v1.6.7 and the best I got was the "ambiguous" behaviour reported by @bennycooly (which I seemingly didn't have before):

```shell
$ dig @127.0.0.1 -x 10.150.206.170

; <<>> DiG 9.10.6 <<>> @127.0.0.1 -x 10.150.206.170
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 34581
;; flags: qr aa rd; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 1
;; WARNING: recursion requested but not available

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
;; QUESTION SECTION:
;170.206.150.10.in-addr.arpa.   IN      PTR

;; ANSWER SECTION:
170.206.150.10.in-addr.arpa. 1  IN      PTR     cilium-etcd-hrb8cstc6f.cilium-etcd-client.cilium.svc.cluster.local.

;; Query time: 0 msec
;; SERVER: 127.0.0.1#53(127.0.0.1)
;; WHEN: Thu Feb 20 08:18:44 WET 2020
;; MSG SIZE  rcvd: 163
```

```shell
$ dig @127.0.0.1 -x 10.150.206.170
; <<>> DiG 9.10.6 <<>> @127.0.0.1 -x 10.150.206.170
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 2178
;; flags: qr aa rd; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 1
;; WARNING: recursion requested but not available

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
;; QUESTION SECTION:
;170.206.150.10.in-addr.arpa.   IN      PTR

;; ANSWER SECTION:
170.206.150.10.in-addr.arpa. 5  IN      PTR     cilium-etcd-hrb8cstc6f.cilium-etcd.cilium.svc.cluster.local.

;; Query time: 0 msec
;; SERVER: 127.0.0.1#53(127.0.0.1)
;; WHEN: Thu Feb 20 08:18:44 WET 2020
;; MSG SIZE  rcvd: 156
```

**Comment by chrisohaver** [comment]
Thanks! I had not spotted that difference in the prior example ... `cilium-etcd` vs `cilium-etcd-client`... definitely two different services.

Fix in progress...

**Comment by bmcustodio** [comment]
@chrisohaver awesome, thank you very much for having a look!

**Comment by chrisohaver** [comment]
#3687 will allow the reverse lookup to reply with all records for the IP.  Such that, say if there are 4 headless services pointing to an endpoint, a reverse lookup would answer with 4 records.  How exactly is the client using the data returned by this reverse lookup?

Will this cause confusion on the client's part?  Receiving multiple answers?  If so, how would we know which answer to provide?  We could prioritize headless services, but what if there are more than one headless service pointing to the ip?

**Comment by chrisohaver** [comment]
> Is this the expected behavior? It seems that CoreDNS isn't doing anything wrong by creating an additional A record with the dashed-ip form even for non-headless services that points to the endpoint IP, but this results in reverse lookups being ambiguous.

It is intentional for endpoint records to exist for non headless services.  It's not required by the spec (nor forbidden).  The ambiguity was caused by CoreDNS arbitrarily selecting the first match returned by the reverse index.

> Similarly, the etcd-operator is correctly setting the hostname for the pods and expects all reverse lookups for the pod IP to be that hostname, not the dashed-ip form of the endpoint.

CoreDNS pulls the hostname from the Endpoints API, which should be populated (by Kubernetes) with the Pod's hostname for each Endpoint.  If CoreDNS is displaying a dashed-ip name, it's because the hostname is empty for the Endpoint.

**Comment by benfuu** [comment]
From the [etcd listener here](https://github.com/etcd-io/etcd/blob/master/pkg/transport/listener_tls.go#L233-L249) it looks like it will iterate over all of the returned PTR records.

**Comment by chrisohaver** [comment]
IIUC, this is fixed with #3687.  Please reopen if you find this does not resolve this issue.
